### PR TITLE
Introduce UseTypedReturnForNewMethodsRule

### PR DESCRIPTION
**Pull request by matks**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | Introduce UseTypedReturnForNewMethodsRule. This rule asserts than any new class method must declare a return type. Exclusions to this rule can be provided through a configuration array, similarly to UseStrictTypesForNewClassesRule.<br/><br/>Also UseTypedReturnForNewMethodsRule looks for inheritance: if class `A` extends `B` and `B::bar()` is part of the exclusion list, then `A::bar()` will be excluded too.
| Type?             | new feature
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | 1 item of https://github.com/PrestaShop/PrestaShop/issues/22728
| How to test?      | CI should pass
| Possible impacts? |

### CODE REVIEW

File: tests/data/MethodBarEWithoutReturnType.php [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Space;
+
+class E
+{
+    public function bar()
```
Comment (matks) [inline tests/data/MethodBarEWithoutReturnType.php]: This class should be refused with the Rule, but it was excluded in configuration from the analysis.

File: README.md [hunk 2]
```diff
@@ -47,9 +47,18 @@ includes:
     - vendor/prestashop/phpstan-prestashop/extension.neon
 ```
 
-You need to provide a service that is an instance of `PHPStanForPrestaShop\PHPConfigurationLoader\ConfigurationLoaderInterface` named `@strictTypesForNewClassesRuleConfigurationFileLoader`. It is used by rule `UseStrictTypesForNewClassesRule`.
+### Required settings
```
Comment (matks) [inline README.md:50]: Rework of the README to make it clearer - see https://github.com/PrestaShop/phpstan-prestashop/blob/ec12ca64f920fdd8ad931fa26e0e5049e4fc5e11/README.md

**Comment by matks** [comment]
Note: looks like I will have to create another rule for functions return types (outside of a class). Class methods and Functions are different entities in the PHP parser library. I should also check anonymous functions.

File: tests/data/MethodBarEWithoutReturnType.php (hunk 1 above)
Comment (matks) [reply tests/data/MethodBarEWithoutReturnType.php]: So it's the usecase "this should rejected by PHPstan but as it was already there before and we dont want to introduce a BC break we ignore it"

**Comment by PierreRambaud** [comment]
Isn't something already present here: https://github.com/phpstan/phpstan-strict-rules ? :thinking:

**Comment by matks** [comment]
> https://github.com/phpstan/phpstan-strict-rules

I did not find such a rule. However even if it existed, I dont think it would allow us to provide an "excluded list" that we can use to avoid phpstan asking us to add BC breaks?

**Comment by matks** [comment]
Woups! Just found out that this Rule cannot be complete unless it also looks for _parent methods in parent classes_.

If class A with function `foo($a)` already exists and is being excluded from analysis because it was there before, and we create a class B that extends A ; then `B::foo($a)` should not be flagged by phpstan as invalid. This rule (in the current implementation) will.

So I add "WIP" to the PR 😅

**Comment by matks** [comment]
Note from @PierreRambaud : check what happens with return type void

**Comment by matks** [comment]
> Woups! Just found out that this Rule cannot be complete unless it also looks for _parent methods in parent classes_.
> 
> If class A with function `foo($a)` already exists and is being excluded from analysis because it was there before, and we create a class B that extends A ; then `B::foo($a)` should not be flagged by phpstan as invalid. This rule (in the current implementation) will.

Implemented in https://github.com/PrestaShop/phpstan-prestashop/pull/9/commits/f7d13f23d836204b20756f077dbd8ae33bce7978 !

But now I see the limits of my `tests/data` folder with wrong namespace, so I will need to clean it in another PR.

**Comment by matks** [comment]
Need https://github.com/PrestaShop/phpstan-prestashop/pull/10 to be merged

**Comment by PierreRambaud** [comment]
> Need #10 to be merged

It has been merged :)

File: src/Rules/UseTypedReturnForNewMethodsRule.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    /**
+     * @param ClassMethod $node
+     * @param Scope $scope
+     *
+     * @return array
+     * @throws \PHPStan\ShouldNotHappenException
+     */
+    public function processNode(Node $node, Scope $scope): array
```
Comment (matks) [inline src/Rules/UseTypedReturnForNewMethodsRule.php:51]: The logic here follows this line:
1. check if function is part of the exclusion list, if yes, return nothing.
2. check if function has return type, if yes, return nothing.
3. if class has parents, look at them and see if they have the same method ; if no, then it's a rule violation.
4. if class has parents who also have this class method, see if they are part of the exclusion list, if yes, return nothing.
5. return violation else.

**Comment by matks** [comment]
> Note from @PierreRambaud : check what happens with return type void

Test case added in https://github.com/PrestaShop/phpstan-prestashop/pull/9/commits/ed2190daa8f46eb097c0786f1c89a9f1db68db06

**Review by Progi1984 — APPROVED** [review]

**Comment by matks** [comment]
Thank you for review @Progi1984 !

**Pull request merged**